src/main/scala/calc/Tree.scala

```diff
@@ -11,16 +16,42 @@ final case class Literal(value: Double)(implicit val pos: Position) extends Tree
 final case class Ident(name: String)(implicit val pos: Position) extends Tree
 
 final case class BinaryOp(op: String, lhs: Tree, rhs: Tree)(
-    implicit val pos: Position) extends Tree
+  implicit val pos: Position) extends Tree
 
 final case class Let(name: Ident, value: Tree, body: Tree)(
-    implicit val pos: Position) extends Tree
+  implicit val pos: Position) extends Tree
 
 final case class Closure(params: List[Ident], body: Tree)(
-    implicit val pos: Position) extends Tree
+  implicit val pos: Position) extends Tree
 
 final case class Call(fun: Tree, args: List[Tree])(
-    implicit val pos: Position) extends Tree
+  implicit val pos: Position) extends Tree
 
 final case class If(cond: Tree, thenp: Tree, elsep: Tree)(
-    implicit val pos: Position) extends Tree
+  implicit val pos: Position) extends Tree
+
+sealed abstract class TreeT {
+  def pos: Position
+  def getType(): Type
```

Comment:
That's a bit Java-ish ;) In Scala we don't use `getX`-style getters. We just use `def tpe: Type` (without `get` nor `()`) or possibly `val tpe: Type`.